transport_plugins/bled112/iotile_transport_bled112/async_packet.py

```diff
@@ -51,37 +60,39 @@ def read_packet(self, timeout=3.0):
             raise InternalTimeoutError("Timeout waiting for packet in AsyncPacketBuffer")
 
 
-def ReaderThread(filelike, read_queue, header_length, length_function, stop):
+def reader_thread(filelike, read_queue, header_length, length_function, stop, dedupe=False, dedupe_timeout=5):
     logger = logging.getLogger(__name__)
+    broadcast_v2_dedupers = None
+    read_buffer = bytearray()
+
+    if dedupe:
+        broadcast_v2_dedupers = BroadcastV2DeduperCollection(dedupe_timeout)
 
     while not stop.is_set():
         try:
-            header = bytearray()
-            while len(header) < header_length:
-                chunk = bytearray(filelike.read(header_length - len(header)))
-                header += chunk
+            # The bled112 will read EOF when there is no more data, so it's safe to read a large amount
+            read_buffer += bytearray(filelike.read(1024))
 
-                if stop.is_set():
+            while not stop.is_set() and len(read_buffer) >= header_length:
+                next_packet_len = header_length + length_function(read_buffer[:header_length])
+
+                if len(read_buffer) < next_packet_len:
+                    # Still waiting to read this packet
                     break
 
-            if stop.is_set():
-                break
+                # Process the packet and remove it from the read buffer
+                packet = read_buffer[:next_packet_len]
+                del read_buffer[:next_packet_len]
 
-            remaining_length = length_function(header)
+                if broadcast_v2_dedupers:
```

Comment:
This can be expressed more simply as:
```python
if broadcsat_v2_dedupers is not None and broadcast_v2_dedupers.allow_packet(packet):
    continue
```

python gurantees left-to-right evaluation of conditions so this is safe and `broadcast_v2_dedupers` will not be accessed if its None.

Reply:
True. Isn't `is not None` the same as `if broadcast_v2_dedupers` though?

Reply:
> Isn't `is not None` the same as `if broadcast_v2_dedupers` though?

More or less.  Technically the first check just if it's assigned explicitly to `None` whereas the second allows any value that evaluates to True, which can lead to some surprises but in this case they're equivalent.  

I personally like the explicitness of comparing to `None` when that's the intent but people differ on that one.

The comment was mainly just about combining the two `if` statements together.

Reply:
Sure, I've already adjusted the if statement. Thanks for the clarification, having the more explicit check is definitely better.